src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/18.diff.sql

```diff
@@ -0,0 +1,297 @@
+        
+        -- Finds the highest boundary value.        
+        DECLARE @startingRightPartitionBoundary datetime2(7) = CAST((SELECT TOP (1) value
+                            FROM sys.partition_range_values AS prv
+                                JOIN sys.partition_functions AS pf
+                                    ON pf.function_id = prv.function_id
+                            WHERE  pf.name = N'PartitionFunction_ResourceChangeData_Timestamp'
+                            ORDER  BY prv.boundary_id DESC) AS datetime2(7));
+                            
+        -- Adds one due to starting from the current hour.
+        DECLARE @numberOfPartitionsToAdd int = @numberOfFuturePartitionsToAdd + 1;    
+        
+        WHILE @numberOfPartitionsToAdd > 0 
+        BEGIN  
+            -- Checks if a partition exists.
+            IF (@startingRightPartitionBoundary < @partitionBoundary) 
+            BEGIN
+                -- Creates new empty partition by creating new boundary value and specifying NEXT USED file group.
+                ALTER PARTITION SCHEME PartitionScheme_ResourceChangeData_Timestamp NEXT USED [PRIMARY];
+                ALTER PARTITION FUNCTION PartitionFunction_ResourceChangeData_Timestamp() SPLIT RANGE(@partitionBoundary); 
+            END;
+                
+            -- Adds one hour for the next partition.
+            SET @partitionBoundary = DATEADD(hour, 1, @partitionBoundary);
+            SET @numberOfPartitionsToAdd -= 1;                 
+        END;
+
+    COMMIT TRANSACTION
+END;
+GO
+
+-- STORED PROCEDURE
+--     RemovePartitionFromResourceChanges
+--
+-- DESCRIPTION
+--     Switches out and merges the extreme left partition with the imediate left partition.
+--     After that, truncates the staging table to purge the old resource change data.
+--
+--     @partitionBoundary
+--         * The output parameter to stores the removed partition boundary.
+--
+CREATE OR ALTER PROCEDURE dbo.RemovePartitionFromResourceChanges
+    @partitionBoundary datetime2(7) OUTPUT
+AS
+  BEGIN
+    
+    --using XACT_ABORT to force a rollback on any error.
+    SET XACT_ABORT ON;
+    
+    BEGIN TRANSACTION
+    
+        -- Finds the lowest boundary value.
+        DECLARE @leftPartitionBoundary datetime2(7) = CAST((SELECT TOP (1) value
+                            FROM sys.partition_range_values AS prv
+                                JOIN sys.partition_functions AS pf
+                                    ON pf.function_id = prv.function_id
+                            WHERE pf.name = N'PartitionFunction_ResourceChangeData_Timestamp'
+                            ORDER BY prv.boundary_id ASC) AS datetime2(7));
+
+        -- Cleans up a staging table if there are existing rows.
+        TRUNCATE TABLE dbo.ResourceChangeDataStaging;
+        
+        -- Switches a partition to the staging table.
+        ALTER TABLE dbo.ResourceChangeData SWITCH PARTITION 2 TO dbo.ResourceChangeDataStaging;
```

Comment:
is PARTITION 2 guaranteed to exist?

I know it is not our use case but for example if EXEC dbo.ConfigurePartitionOnResourceChanges @numberOfFuturePartitionsToAdd = 1

Reply:
Yes, it is guaranteed PARTITION 2 exists. There are no PARTITION 2 means that all partition boundaries were deleted intentionally. In this case, SQL Server should throw the exception.